aspnetcore/security/ip-safelist.md

```diff
@@ -1,70 +1,148 @@
 
-In each case, a string containing approved client IP addresses is stored in an app setting. The middleware or filter parses the string into a list and checks if the remote IP is in the list. If not, an HTTP 403 Forbidden status code is returned.
+In each case, a string containing approved client IP addresses is stored in an app setting. The middleware or filter:
 
-[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/security/ip-safelist/samples/2.x/ClientIpAspNetCore) ([how to download](xref:index#how-to-download-a-sample))
+* Parses the string into an array. 
+* Checks if the remote IP address exists in the array.
 
-## The safelist
+Access is allowed if the array contains the IP address. Otherwise, an HTTP 403 Forbidden status code is returned.
 
-The list is configured in the *appsettings.json* file. It's a semicolon-delimited list and can contain IPv4 and IPv6 addresses.
+[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/security/ip-safelist/samples) ([how to download](xref:index#how-to-download-a-sample))
+
+## IP address safelist
+
+In the sample app, the IP address safelist is:
+
+* Defined by the `AdminSafeList` property in the *appsettings.json* file.
+* A semicolon-delimited string that may contain both [Internet Protocol version 4 (IPv4)](https://wikipedia.org/wiki/IPv4) and [Internet Protocol version 6 (IPv6)](https://wikipedia.org/wiki/IPv6) addresses.
+
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-json[](ip-safelist/samples/3.x/ClientIpAspNetCore/appsettings.json?highlight=2)]
+
+::: moniker-end
+
+::: moniker range="<= aspnetcore-2.2"
 
 [!code-json[](ip-safelist/samples/2.x/ClientIpAspNetCore/appsettings.json?highlight=2)]
 
+::: moniker-end
+
+In the preceding example, the IPv4 addresses of `127.0.0.1` and `192.168.1.5` and the IPv6 loopback address of `::1` (compressed format for `0:0:0:0:0:0:0:1`) are allowed.
+
 ## Middleware
 
-The `Configure` method adds the middleware and passes the safelist string to it in a constructor parameter.
+The `Startup.Configure` method adds the custom `AdminSafeListMiddleware` middleware type to the app's request pipeline. The safelist is retrieved with the .NET Core configuration provider and is passed as a constructor parameter.
 
-[!code-csharp[](ip-safelist/samples/2.x/ClientIpAspNetCore/Startup.cs?name=snippet_Configure&highlight=10)]
+::: moniker range=">= aspnetcore-3.0"
 
-The middleware parses the string into an array and looks for the remote IP address in the array. If the remote IP address is not found, the middleware returns HTTP 401 Forbidden. This validation process is bypassed for HTTP Get requests.
+[!code-csharp[](ip-safelist/samples/3.x/ClientIpAspNetCore/Startup.cs?name=snippet_ConfigureAddMiddleware)]
 
-[!code-csharp[](ip-safelist/samples/2.x/ClientIpAspNetCore/AdminSafeListMiddleware.cs?name=snippet_ClassOnly)]
+::: moniker-end
```

Comment:
10 moniker pairs are too many. Any reason you didn't do the whole moniker versioning we usually use?

Updating to 5 and 5.1 is going to be more difficult.

Reply:
There are 2 LTS versions of the product: 2.1 and 3.1. We need to have docs for both. I could make this doc available only for 2.1+. Thoughts on that approach?

Reply:
This is what we've been doing:

```
::: moniker range=">= aspnetcore-3.0"
// 3.1
::: moniker-end

::: moniker range="<= aspnetcore-3.0"
// 2.1
::: moniker-end
```

It's pretty easy to put it in that format and easy to test.

Reply:
At this point, we'd consolidate the monikers in another PR. This is how we've been working to update docs to 3.1. Update the doc with 2 sections, but don't update the content:

```
::: moniker range=">= aspnetcore-3.0"
//  copy of 2.1, possibly with consolidation
::: moniker-end

::: moniker range="<= aspnetcore-3.0"
// original 2.1
::: moniker-end
```

Then merge the PR into master/live.  The next PR you update the 3.1 section so the diff is easier to follow for the dev team.

Reply:
I've reduced the number of moniker pairs to 4.

We should discuss the versioning strategy offline. Duplicating content in the manner you've proposed above seems like a bad idea.